homeassistant/components/smartthings/__init__.py

```diff
@@ -0,0 +1,201 @@
+
+from aiohttp.client_exceptions import (
+    ClientConnectionError, ClientResponseError)
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_ACCESS_TOKEN
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.dispatcher import (
+    async_dispatcher_connect, async_dispatcher_send)
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType
+
+from .config_flow import SmartThingsFlowHandler  # noqa
+from .const import (
+    CONF_APP_ID, CONF_INSTALLED_APP_ID, DATA_BROKERS, DATA_MANAGER, DOMAIN,
+    SIGNAL_SMARTTHINGS_UPDATE, SUPPORTED_PLATFORMS)
+from .smartapp import (
+    setup_smartapp, setup_smartapp_endpoint, validate_installed_app)
+
+REQUIREMENTS = ['pysmartapp==0.3.0', 'pysmartthings==0.4.2']
+DEPENDENCIES = ['webhook']
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass: HomeAssistantType, config: ConfigType):
+    """Initialize the SmartThings platform."""
+    await setup_smartapp_endpoint(hass)
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry):
+    """Initialize config entry which represents an installed SmartApp."""
+    from pysmartthings import SmartThings
+
+    if not hass.config.api.base_url.lower().startswith('https://'):
+        _LOGGER.warning("The 'base_url' of the 'http' component must be "
+                        "configured and start with 'https://'")
+        return False
+
+    api = SmartThings(async_get_clientsession(hass),
+                      entry.data[CONF_ACCESS_TOKEN])
+
+    remove_entry = False
+    try:
+        # See if the app is already setup. This occurs when there are
+        # installs in multiple SmartThings locations (valid use-case)
+        manager = hass.data[DOMAIN][DATA_MANAGER]
+        smart_app = manager.smartapps.get(entry.data[CONF_APP_ID])
+        if not smart_app:
+            # Validate and setup the app.
+            app = await api.app(entry.data[CONF_APP_ID])
+            smart_app = setup_smartapp(hass, app)
+
+        # Validate and retrieve the installed app.
+        installed_app = await validate_installed_app(
+            api, entry.data[CONF_INSTALLED_APP_ID])
+
+        # Get devices and their current status
+        devices = await api.devices(
+            location_ids=[installed_app.location_id])
+        await asyncio.gather(*[d.status.refresh() for d in devices])
+
+        # Setup device broker
+        broker = DeviceBroker(hass, devices,
+                              installed_app.installed_app_id)
+        broker.event_handler_disconnect = \
+            smart_app.connect_event(broker.event_handler)
+        hass.data[DOMAIN][DATA_BROKERS][entry.entry_id] = broker
+
+    except ClientResponseError as ex:
+        _LOGGER.exception(ex)
+        if ex.status in (400, 401, 403, 422):
```

Comment:
Why is a bad request reason for removal? I can understand 401, unauthorized. 403 could be if we don't have permission maybe to refresh one of the devices?

Reply:
We used to update the SmartApp during setup if the `base_url` was changed, which could result in an unrecoverable 400 or 422 status.  I removed this because I discovered the web server isn't listening during initial HASS setup, but forgot to clean this up. :) .  403 is also unrecoverable as it represents also 'not found', i.e. deleted and the user has to setup again.